codec/src/main/java/io/netty/handler/codec/protobuf/ProtobufVarint32FrameDecoder.java

```diff
@@ -70,52 +70,68 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out)
      *
      * @return decoded int if buffers readerIndex has been forwarded else nonsense value
      */
-    private static int readRawVarint32(ByteBuf buffer) {
+    static int readRawVarint32(ByteBuf buffer) {
+        if (buffer.readableBytes() < 4) {
+            return readRawVarint24(buffer);
+        }
+        int wholeOrMore = buffer.getIntLE(buffer.readerIndex());
+        int firstOneOnStop = ~wholeOrMore & 0x80808080;
+        if (firstOneOnStop == 0) {
+            return readRawVarint40(buffer, wholeOrMore);
+        }
+        int bitsToKeep = Integer.numberOfTrailingZeros(firstOneOnStop) + 1;
+        buffer.skipBytes(bitsToKeep >> 3);
+        int shiftToHighlightTheWhole = 32 - bitsToKeep;
+        // this is not brilliant, we have a data dependency here
+        int wholeWithContinuations = (wholeOrMore << shiftToHighlightTheWhole) >> shiftToHighlightTheWhole;
+        // mix them up as per varint spec while dropping the continuation bits
+        int result = wholeWithContinuations & 0x7F |
```

Comment:
This can be done in two steps instead of four.

```
wwc = (wwc & 0x7F007F) | ((wwc & 0x7F007F00) >> 1);
wwc = (wwc & 0x3FFF) | ((wwc & 0x3FFF0000) >> 2);
```

Probably no difference, just for the kicks.